File: bots/server_module/game_state/available_orders.py
from server_module.game_rules.game_rules import GameRules
from server_module.game_state.order import Order, OrderType
from server_module.game_state.house_type import HouseType
from server_module.game_state.placed_orders import PlacedOrders


class AvailableOrders(dict[HouseType, dict[OrderType, list[Order]]]):
    def __init__(self, **kwargs):
        if kwargs is None:
            super().__init__()
        else:
            for ht, ao in kwargs.items():
                orders_by_type = {}
                for ot in ao:
                    orders = []
                    for i in range(len(ao[ot])):
                        orders.append(Order.from_json(ao[ot][i]))
                    orders_by_type[OrderType.from_str(ot)] = orders
                self[HouseType[ht.upper()]] = orders_by_type

    def build_from_placed_orders(self, game_rules: GameRules, placed_orders: PlacedOrders):
        for ht in HouseType:
            if ht is not HouseType.NEUTRAL:
                self[ht] = {}
                for ot, orders in game_rules.loaded_orders.items():
                    self[ht][ot] = []
                    for o in orders:
                        self[ht][ot].append(o)

        for ht in placed_orders:
            for tn, o in placed_orders[ht].items():
                idx = [i for i, j in enumerate(self[ht][o.order_type])
                       if j.modifier == o.modifier and j.is_star == o.is_star]
                if len(idx):
                    del self[ht][o.order_type][idx[0]]

File: game-logic-core/scala3MainProject/src/main/scala/fwc/game/planningPhase/AvailableOrders.scala
package fwc.game.planningPhase

import fwc.game.houses.*
import enrichment.ExtSeq
import fwc.game.board.{TrackType, Tracks}
import fwc.{JsonParsable, JsonSerializable}
import fwc.game.{FWCException, gameRules}
import ujson.Value

import scala.util.boundary
import scala.util.Try

case class AvailableOrders(
                            orders: Map[HouseType, Map[OrderType, Seq[Order]]]
                          ) extends JsonSerializable {

  def isOrderAvailable(house: HouseType, order: Order): Boolean = {
    this.orders(house)(order.orderType)
      .exists(o => o.isStar == order.isStar && o.modifier == order.modifier)
  }

  def useOrder(house: HouseType, order: Order): AvailableOrders = {
    if (!isOrderAvailable(house, order)) {
      throw new FWCException(s"Order \"${order.orderType} ${if order.modifier != 0 then order.modifier else ""}\" is not available")
    }

    val newOrders = orders(house)(order.orderType).deleteFirstMatch(order)

    AvailableOrders(
      orders + (
        house -> (
          orders(house) + (order.orderType -> newOrders)
        )
      )
    )

  }

  def returnOrder(house: HouseType, order: Order): AvailableOrders = {
    AvailableOrders(
      orders + (
        house -> (
          orders(house) + (
            order.orderType -> (
              orders(house)(order.orderType) :+ order
            )
          )
        )
      )
    )
  }

  def disableOrderType(orderType: OrderType): AvailableOrders = {
    if (orderType == OrderType.March)
      throw new FWCException("cannot disable all march orders")
      
    AvailableOrders(
      this.orders.map((ht, orders) => {
        ht -> (orders + (orderType -> Seq()))
      })
    )
  }
  
  def disableMarchPlusOneOrder(): AvailableOrders = {
    AvailableOrders(
      this.orders.map((ht: HouseType, orders: Map[OrderType, Seq[Order]]) => {
        ht -> (
          orders
            + (OrderType.March -> orders(OrderType.March).filter(!_.isStar))
          ) 
      })
    )
  }

  def toJson: ujson.Value = {
    ujson.Obj(
      upickle.core.LinkedHashMap(
        orders.map((houseType, ordersByType: Map[OrderType, Seq[Order]]) => {
          houseType.toString -> ujson.Obj(
            upickle.core.LinkedHashMap(
              ordersByType.map((orderType, ordersSeq: Seq[Order]) => {
                orderType.toString -> ujson.Value(
                  ordersSeq.map(_.toJson)
                )
              })
            )
          )
        })
      )
    )
  }

  def hasAvailableOrders(houseType: HouseType, tracks: Tracks): Boolean = boundary {
    val availableStars = Try(gameRules.kingsCourtStars(tracks(TrackType.Court).indexOf(houseType))).getOrElse(0)
    val starsShouldBeLeft = 5 - availableStars
    
    val ordersLeft = orders(houseType).view.flatten[Order](_._2).foldLeft(0)(
      (acc, cur) =>
        if cur.isStar then acc + 1 else boundary.break(false)
    )
    ordersLeft > starsShouldBeLeft
  }

}

object AvailableOrders extends JsonParsable {
  def initialize(): AvailableOrders = {
    AvailableOrders(
      Map(
        HouseType.Wolf -> gameRules.loadedOrders,
        HouseType.Moose -> gameRules.loadedOrders,
        HouseType.PufferFish -> gameRules.loadedOrders,
        HouseType.Kraken -> gameRules.loadedOrders,
        HouseType.Rose -> gameRules.loadedOrders,
        HouseType.Lion -> gameRules.loadedOrders
      )
    )
  }

  override def fromJson(json: Value): AvailableOrders = {
    AvailableOrders(
      json.obj.map((houseType, ordersByType: Value) => 
        HouseType.fromString(houseType) -> ordersByType.obj.map((orderType, orders: Value) => 
          OrderType.fromString(orderType) -> orders.arr.map(o => Order.fromJson(o)).toSeq
        ).toMap
      ).toMap
    )
  }
}
